Zend/zend.c

```diff
@@ -769,6 +769,28 @@ void zend_register_standard_ini_entries(void) /* {{{ */
 }
 /* }}} */
 
+static void zend_resolve_property_types(void) /* {{{ */
+{
+	zend_class_entry *ce;
+	zend_property_info *prop_info;
+
+	ZEND_HASH_FOREACH_PTR(CG(class_table), ce) {
+		if (UNEXPECTED(ce->type == ZEND_INTERNAL_CLASS)
+		 && UNEXPECTED(ZEND_CLASS_HAS_TYPE_HINTS(ce))) {
+
+			ZEND_HASH_FOREACH_PTR(&ce->properties_info, prop_info) {
+				if (prop_info->type == IS_OBJECT && prop_info->type_name) {
+					zend_class_entry *prop_ce =
+						(zend_class_entry*)zend_hash_find(CG(class_table), prop_info->type_name);
+
+					assert(prop_ce && prop_ce->type == ZEND_INTERNAL_CLASS);
```

Comment:
assert() would prevent starting PHP with wrong type-hints independently of DEBUG/RELEASE build.